### ISSUE TITLE
Mistyping CommandHandler.Create arguments should throw

### ISSUE THREAD

**Issue by christianrondeau** [dotnet/command-line-api#622]
If you create an argument of one type and then specify another in the command handler, it will silently pass null, which makes it hard to catch and then to diagnose.

```csharp
command.AddOption(new Option("--file") { Argument = new Argument<FileInfo>().ExistingOnly() });
// ...
command.Handler = CommandHandler.Create((string file) =>
{
    // ...
});
```

Here the `FileInfo` cannot be cast to `string` and then fails silently.

I would propose strict arguments matching, i.e. if an argument is of the wrong type, missing or extraneous, it would immediately throw. Since this could potentially break usages, if you prefer this could also be an option, like `command.Handler = CommandHandler.Create(...).Strict()`, which would set a flag on the `ICommandHandler`. The `ParseResultVisitor` would then check that there's a 1:1 match between the handler and the options/arguments.

Since I already have a few PRs waiting, I'll wait for feedback before providing a PR, but I'm ready to make one if you believe this makes sense.

**Comment by jonsequitur** [comment]
Is it your expectation that an exception should be thrown when `command.Handler` is assigned to and the `file` parameter's type doesn't match the corresponding `Argument`'s type?

**Comment by christianrondeau** [comment]
That's correct (or coerce into the target type). That can avoid surprises when refactoring names for example.

**Comment by jonsequitur** [comment]
I think that adding a method such as your `Strict` suggestion to do this verification would be very useful. As it will have a non-zero performance cost, making it a separate method would allow it to be used in tests or inside an `#if DEBUG` block to avoid the cost during every app startup.

We would welcome a PR.
